community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureHolder.java

```diff
@@ -53,20 +56,52 @@ T get( int id )
         return store.get( id );
     }
 
-    void put( QualifiedName name, T item )
+    void put( QualifiedName name, T item, boolean caseInsensitive )
     {
         int id = store.size();
         store.add( item );
         nameToId.put( name, id );
+        if ( caseInsensitive )
+        {
+            caseInsensitveName2Id.put( toLowerCaseName( name ), id );
+        }
+        else
+        {
+            caseInsensitveName2Id.remove( toLowerCaseName( name ) );
+        }
     }
 
     int idOf( QualifiedName name )
     {
-        return nameToId.get( name );
+        Integer id = name2Id( name );
+        if ( id == null )
+        {
+            throw new NoSuchElementException();
+        }
+
+        return id;
     }
 
     List<T> all()
     {
         return Collections.unmodifiableList( store );
     }
+
+    private Integer name2Id( QualifiedName name )
+    {
+        Integer id = nameToId.get( name );
+        if ( id == null )
+        { // Did not find it in the case sensitive lookup - let's check for case insensitive objects
+            QualifiedName lowerCaseName = toLowerCaseName( name );
+            id = caseInsensitveName2Id.get( lowerCaseName );
+        }
+        return id;
+    }
+
+    private QualifiedName toLowerCaseName( QualifiedName name )
+    {
+        String[] lowerCaseNamespace = Arrays.stream( name.namespace() ).map( String::toLowerCase ).toArray( String[]::new );
```

Comment:
You can size the array upfront in this case since all sizes are known, in that way you don't need to create as many intermediate objects as in the lambda way.